magma/bits.py

```diff
@@ -72,13 +75,98 @@ class BitsMeta(AbstractBitVectorMeta, ArrayMeta):
     def __new__(mcs, name, bases, namespace, info=(None, None, None), **kwargs):
         return ArrayMeta.__new__(mcs, name, bases, namespace, info, **kwargs)
 
+    def _make_const(cls, value: tp.Union[tuple, int]):
+        """
+        value can be a tuple of bits or an object that supports the `int`
+        function
+        """
+        if isinstance(value, tuple):
+            value = seq2int(value)
+        else:
+            value = int(value)
+        value = cls.hwtypes_T[cls.N](value)
+        return Out(cls)(const_value=value, name=ConstRef(repr(value)))
+
+    def _make_from_int_const(cls, arg: int):
+        if arg.bit_length() > cls.N:
+            raise ValueError(
+                f"Cannot construct {cls.orig_name}[{cls.N}] with "
+                f"integer {arg} (requires truncation)")
+        return cls._make_const(tuple(int2seq(arg, cls.N)))
+
+    def _make_from_bv_const(cls, arg: BitVector):
+        if len(arg) != cls.N:
+            raise TypeError(
+                f"Cannot construct {cls.orig_name}[{cls.N}] with "
+                f"BitVector of length {len(arg)} (sizes must "
+                "match)")
+        return cls._make_const(tuple(arg.bits()))
+
+    def _make_from_bits(cls, arg: 'Bits', kwargs):
+        if arg.const():
+            return cls._make_const(tuple(int2seq(int(arg), cls.N)))
+        arg_len = len(arg)
+        if type(arg) is cls:
+            # Don't need to cast
+            return arg
+        if arg_len > cls.N:
+            raise TypeError(
+                "Will not do implicit truncation of bits length")
+
+        args = arg.ts
+        if arg_len < cls.N:
+            args = cls._extend(args)
+        return super().__call__(*args, **kwargs)
+
+    def _make_from_list(cls, arg: tp.List, kwargs):
+        if len(arg) != len(cls):
+            raise TypeError(
+                f"List initializer for Bits[{len(cls)}] must be same "
+                f"length, not {len(arg)}"
+            )
+
+        def _const(x):
+            if isinstance(x, Type) and x.const():
+                return True
+            if isinstance(x, (bool, int, ht.Bit, BitVector)):
+                return True
+            return False
+        if all(_const(x) for x in arg):
+            return cls._make_const(seq2int(list(int(x) for x in arg)))
+        return super().__call__(arg, **kwargs)
+
+    def _make_from_wireable(cls, arg: Type):
+        # Type conversion done with wiring to an anon value
+        result = cls.undirected_t()
+        result @= arg
+        return result
+
+    def _make_from_bit(cls, arg: Bit, kwargs):
+        if arg.const():
+            return cls._make_const(int(arg))
+        return super().__call__([arg], **kwargs)
+
+    def _make_from_one_arg(cls, arg, kwargs):
```

Comment:
from previous convo:

> It seems a bit odd to make them free if they take cls and now we have to pass it explicitly? I'd normally be fine with free functions when cls is not used, but the implicit cls argument seems warranted here since the logic uses the reference.

Not really a big deal but since there's only one entry point (`_make_from_one_arg`) and the rest are basically passing around `cls` anyway, [`cls.x(y, z)` vs `x(cls, y, z)`] and since `cls` is basically only being used as a container for `(N, T), it seems better to make them free. If we were modifying `cls` I would definitely agree. Not a huge deal, just seems slightly aesthetically better that way.

Reply:
Some of these functions call super(), which feels odd to put not in a class method.

Reply:
ok think I missed that part, good point. fine to leave as is.